File: go/inject_windows.go
//go:build windows
// +build windows

package injgo

import (
	"errors"
	"path/filepath"
	"syscall"
	"unsafe"

	"go.zoe.im/injgo/pkg/w32"
)

// Platform-specific errors
var (
	ErrCreateSnapshot = errors.New("failed to create process snapshot")
	ErrModuleSnapshot = errors.New("failed to create module snapshot")
	ErrModuleNotFound = errors.New("module not found in process")
)

// findProcessByName finds a process by name (Windows implementation)
func findProcessByName(name string) (*Process, error) {
	handle, err := syscall.CreateToolhelp32Snapshot(syscall.TH32CS_SNAPPROCESS, 0)
	if err != nil {
		return nil, ErrCreateSnapshot
	}
	defer syscall.CloseHandle(handle)

	var entry syscall.ProcessEntry32
	entry.Size = uint32(unsafe.Sizeof(entry))

	err = syscall.Process32First(handle, &entry)
	if err != nil {
		return nil, ErrProcessNotFound
	}

	for {
		exeFile := w32.UTF16PtrToString(&entry.ExeFile[0])
		if exeFile == name || filepath.Base(exeFile) == name {
			return &Process{
				PID:     int(entry.ProcessID),
				Name:    exeFile,
				ExePath: exeFile, // TODO: get full path
			}, nil
		}

		err = syscall.Process32Next(handle, &entry)
		if err != nil {
			break
		}
	}

	return nil, ErrProcessNotFound
}

// findProcessByPID opens a process by PID (Windows implementation)
func findProcessByPID(pid int) (*Process, error) {
	handle, err := syscall.OpenProcess(syscall.PROCESS_QUERY_INFORMATION, false, uint32(pid))
	if err != nil {
		return nil, ErrProcessNotFound
	}
	syscall.CloseHandle(handle)

	// Get process name
	name := ""
	snapshot, _ := syscall.CreateToolhelp32Snapshot(syscall.TH32CS_SNAPPROCESS, 0)
	if snapshot != 0 {
		defer syscall.CloseHandle(snapshot)
		var entry syscall.ProcessEntry32
		entry.Size = uint32(unsafe.Sizeof(entry))
		syscall.Process32First(snapshot, &entry)
		for {
			if int(entry.ProcessID) == pid {
				name = w32.UTF16PtrToString(&entry.ExeFile[0])
				break
			}
			if syscall.Process32Next(snapshot, &entry) != nil {
				break
			}
		}
	}

	return &Process{
		PID:  pid,
		Name: name,
	}, nil
}

// findAllProcessesByName finds all processes matching the name
func findAllProcessesByName(name string) ([]*Process, error) {
	handle, err := syscall.CreateToolhelp32Snapshot(syscall.TH32CS_SNAPPROCESS, 0)
	if err != nil {
		return nil, ErrCreateSnapshot
	}
	defer syscall.CloseHandle(handle)

	var processes []*Process
	var entry syscall.ProcessEntry32
	entry.Size = uint32(unsafe.Sizeof(entry))

	err = syscall.Process32First(handle, &entry)
	if err != nil {
		return processes, nil
	}

	for {
		exeFile := w32.UTF16PtrToString(&entry.ExeFile[0])
		if name == "" || exeFile == name || filepath.Base(exeFile) == name {
			processes = append(processes, &Process{
				PID:     int(entry.ProcessID),
				Name:    exeFile,
				ExePath: exeFile,
			})
		}

		err = syscall.Process32Next(handle, &entry)
		if err != nil {
			break
		}
	}

	return processes, nil
}

// isInjected checks if a DLL is loaded in the process
func isInjected(pid int, lib string) (bool, error) {
	dllName := filepath.Base(lib)

	handle := w32.CreateToolhelp32Snapshot(w32.TH32CS_SNAPMODULE|w32.TH32CS_SNAPMODULE32, uint32(pid))
	if handle == 0 {
		return false, nil // Can't check, assume not injected
	}
	defer w32.CloseHandle(handle)

	var entry w32.MODULEENTRY32
	entry.Size = uint32(unsafe.Sizeof(entry))

	if !w32.Module32First(handle, &entry) {
		return false, nil
	}

	for {
		moduleName := w32.UTF16PtrToString(&entry.SzModule[0])
		if moduleName == dllName {
			return true, nil
		}
		if !w32.Module32Next(handle, &entry) {
			break
		}
	}

	return false, nil
}

// inject loads a DLL into the target process
func inject(pid int, lib string) error {
	// Get absolute path
	absPath, err := filepath.Abs(lib)
	if err != nil {
		return ErrLibraryNotFound
	}

	// Open process
	handle, err := w32.OpenProcess(w32.PROCESS_ALL_ACCESS, false, uint32(pid))
	if err != nil {
		return err
	}
	defer w32.CloseHandle(handle)

	// Allocate memory for DLL path
	pathLen := len(absPath) + 1
	remoteAddr, err := w32.VirtualAllocEx(handle, 0, pathLen, w32.MEM_COMMIT, w32.PAGE_EXECUTE_READWRITE)
	if err != nil {
		return err
	}

	// Write DLL path
	err = w32.WriteProcessMemory(handle, uint32(remoteAddr), []byte(absPath), uint(pathLen))
	if err != nil {
		return err
	}

	// Get LoadLibraryA address
	kernel32 := w32.GetModuleHandleA("kernel32.dll")
	loadLibAddr, err := w32.GetProcAddress(kernel32, "LoadLibraryA")
	if err != nil {
		return err
	}

	// Create remote thread
	threadHandle, _, err := w32.CreateRemoteThread(handle, nil, 0, uint32(loadLibAddr), remoteAddr, 0)
	if err != nil {
		return err
	}
	defer w32.CloseHandle(threadHandle)

	// Wait for completion (with timeout)
	w32.WaitForSingleObject(threadHandle, 5000)

	return nil
}

// eject unloads a DLL from the target process
func eject(pid int, lib string) error {
	dllName := filepath.Base(lib)

	// Find module handle
	handle := w32.CreateToolhelp32Snapshot(w32.TH32CS_SNAPMODULE|w32.TH32CS_SNAPMODULE32, uint32(pid))
	if handle == 0 {
		return ErrModuleSnapshot
	}
	defer w32.CloseHandle(handle)

	var entry w32.MODULEENTRY32
	entry.Size = uint32(unsafe.Sizeof(entry))
	var moduleHandle uintptr

	if w32.Module32First(handle, &entry) {
		for {
			moduleName := w32.UTF16PtrToString(&entry.SzModule[0])
			if moduleName == dllName {
				moduleHandle = uintptr(unsafe.Pointer(entry.ModBaseAddr))
				break
			}
			if !w32.Module32Next(handle, &entry) {
				break
			}
		}
	}

	if moduleHandle == 0 {
		return ErrModuleNotFound
	}

	// Open process
	procHandle, err := w32.OpenProcess(w32.PROCESS_ALL_ACCESS, false, uint32(pid))
	if err != nil {
		return err
	}
	defer w32.CloseHandle(procHandle)

	// Get FreeLibrary address
	kernel32 := w32.GetModuleHandleA("kernel32.dll")
	freeLibAddr, err := w32.GetProcAddress(kernel32, "FreeLibrary")
	if err != nil {
		return err
	}

	// Create remote thread to call FreeLibrary
	threadHandle, _, err := w32.CreateRemoteThread(procHandle, nil, 0, uint32(freeLibAddr), moduleHandle, 0)
	if err != nil {
		return err
	}
	defer w32.CloseHandle(threadHandle)

	w32.WaitForSingleObject(threadHandle, 5000)

	return nil
}

File: rust/src/inject_windows.rs
//! Windows DLL injection implementation
//!
//! Uses CreateRemoteThread + LoadLibraryW technique.

use std::mem;
use std::path::Path;
use std::ptr::null_mut;

use crate::error::{Error, Result};
use crate::inject::Injector;
use crate::process_windows::Process;
use crate::utils::to_wide_string;
use crate::winapi::*;

impl Injector for Process {
    fn is_injected(&self, dll: &str) -> Result<bool> {
        // Get module name from path
        let dll_name = Path::new(dll)
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or(dll)
            .to_lowercase();

        // Enumerate modules
        let handle = unsafe { 
            CreateToolhelp32Snapshot(TH32CS_SNAPMODULE | TH32CS_SNAPMODULE32, self.pid()) 
        };
        
        if handle.is_null() || handle == INVALID_HANDLE_VALUE {
            // Access denied or process not accessible
            return Ok(false);
        }

        let mut entry: MODULEENTRY32 = unsafe { std::mem::zeroed() };
        entry.dwSize = mem::size_of::<MODULEENTRY32>() as u32;

        let mut found = false;
        
        if unsafe { Module32First(handle, &mut entry) } != 0 {
            loop {
                let module_name = char_to_string(&entry.szModule).to_lowercase();
                if module_name.contains(&dll_name) || dll_name.contains(&module_name) {
                    found = true;
                    break;
                }
                
                if unsafe { Module32Next(handle, &mut entry) } == 0 {
                    break;
                }
            }
        }

        unsafe { CloseHandle(handle) };
        Ok(found)
    }

    fn inject(&self, dll: &str) -> Result<()> {
        // Resolve full path
        let fullpath = Path::new(dll)
            .canonicalize()
            .map_err(|e| Error::library_not_found(format!("{}: {}", dll, e)))?;
        
        let dll_path = fullpath
            .to_str()
            .ok_or_else(|| Error::invalid_argument("invalid path encoding"))?;

        // Convert to wide string
        let path_wstr = to_wide_string(dll_path);
        let path_len = path_wstr.len() * 2;

        // Allocate memory in target process
        let remote_path = unsafe {
            VirtualAllocEx(
                self.handle(),
                null_mut(),
                path_len,
                MEM_RESERVE | MEM_COMMIT,
                PAGE_READWRITE,
            )
        };

        if remote_path.is_null() {
            return Err(Error::platform(
                std::io::Error::last_os_error().raw_os_error().unwrap_or(0),
                "failed to allocate memory in target process",
            ));
        }

        // Write DLL path to target process
        let written = unsafe {
            WriteProcessMemory(
                self.handle(),
                remote_path,
                path_wstr.as_ptr() as _,
                path_len,
                null_mut(),
            )
        };

        if written == FALSE {
            unsafe { VirtualFreeEx(self.handle(), remote_path, 0, MEM_RELEASE) };
            return Err(Error::platform(
                std::io::Error::last_os_error().raw_os_error().unwrap_or(0),
                "failed to write path to target memory",
            ));
        }

        // Get LoadLibraryW address
        let load_library = unsafe {
            GetProcAddress(
                GetModuleHandleA("kernel32.dll\0".as_ptr() as _),
                "LoadLibraryW\0".as_ptr() as _,
            )
        };

        if load_library.is_null() {
            unsafe { VirtualFreeEx(self.handle(), remote_path, 0, MEM_RELEASE) };
            return Err(Error::platform(0, "failed to get LoadLibraryW address"));
        }

        // Create remote thread to call LoadLibraryW
        let thread = unsafe {
            CreateRemoteThread(
                self.handle(),
                null_mut(),
                0,
                Some(mem::transmute(load_library)),
                remote_path,
                0,
                null_mut(),
            )
        };

        if thread.is_null() {
            unsafe { VirtualFreeEx(self.handle(), remote_path, 0, MEM_RELEASE) };
            return Err(Error::from_win32());
        }

        // Wait for thread completion
        let wait_result = unsafe { WaitForSingleObject(thread, 5000) }; // 5 second timeout
        
        if wait_result == WAIT_FAILED {
            unsafe {
                CloseHandle(thread);
                VirtualFreeEx(self.handle(), remote_path, 0, MEM_RELEASE);
            }
            return Err(Error::from_win32());
        }

        // Get thread exit code (LoadLibrary return value)
        let mut exit_code: u32 = 0;
        unsafe {
            GetExitCodeThread(thread, &mut exit_code as *mut _ as _);
            CloseHandle(thread);
            VirtualFreeEx(self.handle(), remote_path, 0, MEM_RELEASE);
        }

        // exit_code == 0 means LoadLibrary failed
        if exit_code == 0 {
            return Err(Error::platform(0, "LoadLibrary returned NULL - DLL load failed"));
        }

        Ok(())
    }

    fn eject(&self, dll: &str) -> Result<()> {
        // Get module name from path
        let dll_name = Path::new(dll)
            .file_name()
            .and_then(|n| n.to_str())
            .unwrap_or(dll)
            .to_lowercase();

        // Find module handle
        let handle = unsafe { 
            CreateToolhelp32Snapshot(TH32CS_SNAPMODULE | TH32CS_SNAPMODULE32, self.pid()) 
        };
        
        if handle.is_null() || handle == INVALID_HANDLE_VALUE {
            return Err(Error::from_win32());
        }

        let mut entry: MODULEENTRY32 = unsafe { std::mem::zeroed() };
        entry.dwSize = mem::size_of::<MODULEENTRY32>() as u32;

        let mut module_handle: Option<HMODULE> = None;
        
        if unsafe { Module32First(handle, &mut entry) } != 0 {
            loop {
                let module_name = char_to_string(&entry.szModule).to_lowercase();
                if module_name.contains(&dll_name) || dll_name.contains(&module_name) {
                    module_handle = Some(entry.hModule);
                    break;
                }
                
                if unsafe { Module32Next(handle, &mut entry) } == 0 {
                    break;
                }
            }
        }

        unsafe { CloseHandle(handle) };

        let hmodule = module_handle
            .ok_or_else(|| Error::library_not_found(format!("{} not loaded", dll)))?;

        // Get FreeLibrary address
        let free_library = unsafe {
            GetProcAddress(
                GetModuleHandleA("kernel32.dll\0".as_ptr() as _),
                "FreeLibrary\0".as_ptr() as _,
            )
        };

        if free_library.is_null() {
            return Err(Error::platform(0, "failed to get FreeLibrary address"));
        }

        // Create remote thread to call FreeLibrary
        let thread = unsafe {
            CreateRemoteThread(
                self.handle(),
                null_mut(),
                0,
                Some(mem::transmute(free_library)),
                hmodule as _,
                0,
                null_mut(),
            )
        };

        if thread.is_null() {
            return Err(Error::from_win32());
        }

        // Wait for completion
        unsafe {
            WaitForSingleObject(thread, 5000);
            CloseHandle(thread);
        }

        Ok(())
    }
}

/// Convert C char array to String
fn char_to_string(chars: &[i8]) -> String {
    chars.iter()
        .map(|c| *c as u8 as char)
        .collect::<String>()
        .trim_end_matches('\0')
        .to_string()
}

// Additional Windows-specific constants
const TH32CS_SNAPMODULE: u32 = 0x00000008;
const TH32CS_SNAPMODULE32: u32 = 0x00000010;

// Module32First/Next bindings
#[link(name = "kernel32")]
extern "system" {
    fn Module32First(hSnapshot: HANDLE, lpme: *mut MODULEENTRY32) -> i32;
    fn Module32Next(hSnapshot: HANDLE, lpme: *mut MODULEENTRY32) -> i32;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_inject_api() {
        // This test requires a running process and DLL
        // Just verify the API compiles
        println!("Inject API test - compile check only");
    }
}
